src/responsibleai/rai_analyse/generate_pdf.py

```diff
@@ -0,0 +1,1226 @@
+
+
+def get_fi_bar_plot(data):
+    y_data = [v["short_label"] for k, v in data.items()]
+    x_data = [v["value"] for k, v in data.items()]
+    max_x = max(x_data)
+    x_range = [0, max_x]
+    x_data = [[x, max_x - x] for x in x_data]
+
+    # tickvals = [0.0, 0.25, 0.5, 0.75, 1.0]
+    # ticktext = [0.0, 0.25, 0.5, 0.75, 1.0]
+    tickappend = ""
+    
+    return get_bar_plot(y_data, x_data, tickappend=tickappend, xrange=x_range)
+    
+
+def get_binary_cp_bar_plot(data, m):
+    metric_name = m
+    y_data = [y["cohort_short_name"] for y in data["cohorts"]]
+    x_data = [int(x[metric_name]*100) for x in data["cohorts"]]
+    x_data = [[x, 100-x] for x in x_data]
+    legend = [m]
+    tickvals = [0, 25, 50, 75, 100]
+    ticktext = [str(x)+"%" for x in tickvals]
+    
+    return get_bar_plot(y_data, x_data, legend=legend,
+                        tickvals=tickvals, ticktext=ticktext, tickappend="%")
+
+
+def get_de_bar_plot(data):
+    y_data = [y["label"] + "<br>" + str(int(y["population"]*100))+"% pop." for y in data["classes"]]
+    x_data = [int(float(x["prediction_0_ratio"])*100) for x in data["classes"]]
+    x_data = [[x, 100-x] for x in x_data]
+    legend = ["Predicted as \"" + y["prediction_0_name"] + "\"" for y in data["classes"]]
+    tickvals = [0, 25, 50, 75, 100]
+    ticktext = [str(x)+"%" for x in tickvals]
+
+    return get_bar_plot(y_data, x_data, legend=legend, tickvals=tickvals, ticktext=ticktext, tickappend="%")
+
+
+def get_dot_plot(center, ep, em):
+    fig = go.Figure()
+    fig.add_trace(go.Scatter(
+        x=[center], y=[0.5], mode='markers',
+        error_x=dict(
+            type='data',
+            symmetric=False,
+            array=[ep],
+            arrayminus=[em],
+            color="rgba(39, 110, 237, 1)",
+            width=15
+            )
+    ))
+    fig.update_xaxes(showgrid=False,
+                     # tickvals=[0, 25, 50, 75, 100],
+                     # ticktext=["0%", "25%", "50%", "75%", "100%"],
+                     # range=[0, 100]
+                     )
+    fig.update_yaxes(showgrid=False,
+                     zeroline=True, zerolinecolor='black', zerolinewidth=3,
+                     showticklabels=False)
+    fig.update_layout(height=240, width=700, plot_bgcolor='white')
+
+    fig.update_traces(marker=dict(color="white", size=20,
+                                  line=dict(width=2,
+                                            color='rgba(39, 110, 237, 1)')))
+
+    png = pio.to_image(fig)
+    png_base64 = base64.b64encode(png).decode('ascii')
+
+    return png_base64
+
+
+def get_bar_plot(y_data, x_data, legend=None,
+                 threshold=None, tickvals=None, ticktext=None,
+                 xrange=None, tickappend=""):
+    fig = go.Figure()
+    series = 0
+    
+    def get_colors(series, index):
+        colors2 = ['rgba(39, 110, 237, 1)',
+          'rgba(218, 227, 243, 1)']
+        return colors2[index%2]
+
+    def get_show_legend(legend, index):
```

Comment:
I'm not a huge fan of nesting functions inside functions, but that's a minor thing at this point.

Reply:
usually function level import and function declaration is for avoiding pollution in the module scope if the function/module only serve a single purpose. These function scope declared function are sort of like lambda except they have enough logic in there that writing a block of lambda may not be as readable.

But I'm happy to follow code convention here.  Do you have suggestions for this?

Reply:
to be fixed in follow up.